alpha/veneer/semver/semver.go

```diff
@@ -0,0 +1,386 @@
+	if err != nil {
+		return nil, err
+	}
+	isvd[CandidateChannelName] = *bdm
+
+	bdm, err = addBundlesToChannel(sv.Fast.Bundles, cfg)
+	if err != nil {
+		return nil, err
+	}
+	isvd[FastChannelName] = *bdm
+
+	bdm, err = addBundlesToChannel(sv.Stable.Bundles, cfg)
+	if err != nil {
+		return nil, err
+	}
+	isvd[StableChannelName] = *bdm
+
+	return &isvd, nil
+}
+
+func ReadFile(ref string) (*semverVeneer, error) {
+	data, err := ioutil.ReadFile(ref)
+	if err != nil {
+		return nil, err
+	}
+
+	// default behavior is to generate only minor channels and to use skips over replaces
+	sv := semverVeneer{
+		GenerateMajorChannels: false,
+		GenerateMinorChannels: true,
+		AvoidSkipPatch:        false,
+	}
+	if err := yaml.Unmarshal(data, &sv); err != nil {
+		return nil, err
+	}
+	return &sv, nil
+}
+
+func (v Veneer) Render(ctx context.Context) (*declcfg.DeclarativeConfig, error) {
+	var out declcfg.DeclarativeConfig
+	// fmt.Printf("<--> Received config: skip(%t) major(%t) minor(%t) ref(%s)\n", v.SkipPatch, v.ChannelsMajor, v.ChannelsMinor, v.Ref)
+
+	sv, err := ReadFile(v.Ref)
+	if err != nil {
+		log.Fatalf("semver-render: unable to read file: %v", err)
+	}
+	fmt.Printf("Semver-Veneer parsed\n")
+	// sv.write()
+
+	var cfgs []declcfg.DeclarativeConfig
+	for _, b := range sv.Candidate.Bundles {
+		r := action.Render{
+			AllowedRefMask: action.RefBundleImage,
+			Refs:           []string{b.Image},
+		}
+		c, err := r.Run(ctx)
+		if err != nil {
+			return nil, err
+		}
+		cfgs = append(cfgs, *c)
+	}
+	out = *combineConfigs(cfgs)
+
+	fmt.Printf("all bundles parsed\n")
+	for _, b := range out.Bundles {
+		fmt.Printf("{ bundle Name:%q Image:%q }\n", b.Name, b.Image)
+	}
+
+	cdm, err := sv.addBundlesToStandardChannels(&out)
+	if err != nil {
+		log.Fatalf("semver-render: unable to post-process bundle info: %v", err)
+	}
+	// sv.write()
+	// fmt.Printf("<--> <-->\n")
+	// isvd.write()
+	channels := sv.decomposeChannels(cdm)
+	out.Channels = channels
+
+	return &out, nil
+}
+
+func (sv *semverVeneer) addChannels(data map[string][]semver.Version, bpath string) []declcfg.Channel {
+	channels := []declcfg.Channel{}
+	for cvername, versions := range data {
+		c := newChannel(bpath, cvername)
+		// add all (unlinked) entries to channel
+		bpkg := path.Base(bpath)
+		for _, ver := range versions {
+			bname := bpkg + "." + ver.String()
+			c.Entries = append(c.Entries, declcfg.ChannelEntry{Name: bname})
+		}
+
+		// link up the edges according to config
+		if sv.AvoidSkipPatch {
+			for i := 1; i < len(c.Entries); i++ {
+				c.Entries[i] = declcfg.ChannelEntry{
+					Name:     c.Entries[i].Name,
+					Replaces: c.Entries[i-1].Name,
+				}
+			}
+		} else {
+			maxIndex := len(c.Entries) - 1
+			curSkips := sets.NewString()
+			for i := 0; i < maxIndex; i++ {
+				curSkips = curSkips.Insert(c.Entries[i].Name)
+			}
+			c.Entries[maxIndex] = declcfg.ChannelEntry{
+				Name:  c.Entries[maxIndex].Name,
+				Skips: curSkips.List(),
+			}
```

Comment:
This appears to add all non-head entries in the channel as skips of the head. This is undesirable for major version channels because it means we could jump over several minor versions at a time.

For major channels we need:
1. A `replaces` chain connecting the highest patch version of each of the minor versions.
2. Each of the members of the replaces chain from 1 should skip other members of their minor version.

So we need:
```mermaid
graph LR
   1.2.2 -- replaces --> 1.1.3
   1.2.2 -- skips --> 1.2.1
   1.2.2 -- skips --> 1.2.0

   1.1.3 -- replaces --> 1.0.1
   1.1.3 -- skips --> 1.1.2
   1.1.3 -- skips --> 1.1.1
   1.1.3 -- skips --> 1.1.0

   1.0.1 -- skips --> 1.0.0



```

Reply:
I _think_ this code hooked the edges up correctly when I was experimenting with something like this: https://github.com/joelanford/declcfg/blob/d1ebdcf2717649e9b867710a8166b738fa63966e/internal/action/semver.go#L67-L90

Reply:
Updated with this, plus `-o mermaid` mermaid output support so we can visually debug (plus some traditional unit tests ;) ) :

```mermaid
graph LR
testoperator.v0.1.0
testoperator.v0.1.1
testoperator.v0.1.2
testoperator.v0.1.3
testoperator.v0.1.3-- skips --> testoperator.v0.1.0
testoperator.v0.1.3-- skips --> testoperator.v0.1.1
testoperator.v0.1.3-- skips --> testoperator.v0.1.2
testoperator.v0.2.0
testoperator.v0.2.1
testoperator.v0.2.2
testoperator.v0.2.2-- replaces --> testoperator.v0.1.3
testoperator.v0.2.2-- skips --> testoperator.v0.2.0
testoperator.v0.2.2-- skips --> testoperator.v0.2.1
testoperator.v0.3.0
testoperator.v0.3.0-- replaces --> testoperator.v0.2.2
```
This is channel `candidate-v0`